File: app/src/main/java/com/inspiringteam/xchange/data/source/remote/QuakesRemoteDataModule.java
package com.inspiringteam.xchange.data.source.remote;

import android.app.Application;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.google.gson.FieldNamingPolicy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.inspiringteam.xchange.di.scopes.AppScoped;

import dagger.Module;
import dagger.Provides;
import okhttp3.Cache;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

import static com.inspiringteam.xchange.util.Constants.QUAKES_API_BASE_URL;

@Module
public class QuakesRemoteDataModule {

    @AppScoped
    @Provides
    QuakesApiService provideQuakesService(Retrofit retrofit) {
        return retrofit.create(QuakesApiService.class);
    }

    @Provides
    @AppScoped
    Retrofit provideRetrofit(Gson gson) {
        return new Retrofit.Builder()
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .baseUrl(QUAKES_API_BASE_URL)
                .build();
    }

    @Provides
    @AppScoped
    Gson provideGson() {
        GsonBuilder gsonBuilder = new GsonBuilder();
        gsonBuilder.setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES);
        return gsonBuilder.create();
    }
}

File: app/src/main/java/com/inspiringteam/xchange/data/source/remote/QuakesRemoteDataModule.kt
package com.inspiringteam.xchange.data.source.remote

import com.google.gson.FieldNamingPolicy
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.inspiringteam.xchange.di.scopes.AppScoped
import com.inspiringteam.xchange.util.Constants
import dagger.Module
import dagger.Provides
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

@Module
class QuakesRemoteDataModule {
    @AppScoped
    @Provides
    fun provideQuakesService(retrofit: Retrofit): QuakesApiService {
        return retrofit.create(QuakesApiService::class.java)
    }

    @Provides
    @AppScoped
    fun provideRetrofit(gson: Gson): Retrofit {
        return Retrofit.Builder()
            .addConverterFactory(GsonConverterFactory.create(gson))
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .baseUrl(Constants.QUAKES_API_BASE_URL)
            .build()
    }

    @Provides
    @AppScoped
    fun provideGson(): Gson {
        val gsonBuilder = GsonBuilder()
        gsonBuilder.setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
        return gsonBuilder.create()
    }

}